File: example/sigma/plot_meff_vs_rs.jl
using CurveFit
using DelimitedFiles
using Measurements
using PyCall
using PyPlot

@pyimport scienceplots  # `import scienceplots` is required as of version 2.0.0
@pyimport scipy.interpolate as interp

function spline(x, y, e; xmin=x[1], xmax=x[end])
    # generate knots with spline without constraints
    w = 1.0 ./ e
    spl = interp.UnivariateSpline(x, y; w=w, k=3)
    __x = collect(LinRange(xmin, xmax, 100))
    yfit = spl(__x)
    return __x, yfit
end

function spline_with_bc(x, y, e; xmin=0.0, xmax=x[end])
    _x, _y = deepcopy(x), deepcopy(y)
    _w = 1.0 ./ e

    #enforce left boundary condition: m*/m → 1 as rs → 0
    rescale = 10000
    pushfirst!(_x, 0.0)
    pushfirst!(_y, 1.0)
    pushfirst!(_w, _w[1] * rescale)

    # generate knots with spline without constraints
    spl = interp.UnivariateSpline(_x, _y; w=_w, k=3)
    __x = collect(LinRange(0.0, x[end], 1000))
    yfit = spl(__x)
    return __x, yfit
end

dim = 3
spin = 2
Fs = -0.0
beta = 40.0
maxOrder = 5
rslist = [0.5, 1.0, 2.0, 3.0, 4.0, 5.0]

cdict = Dict(["blue" => "#0077BB", "cyan" => "#33BBEE", "teal" => "#009988", "orange" => "#EE7733", "red" => "#CC3311", "magenta" => "#EE3377", "grey" => "#BBBBBB"]);

# meff_data = Dict(
#     1.0 => [0.93574 ± 0.0001, 0.94731 ± 0.00011, 0.94992 ± 0.00013, 0.95129 ± 0.00025, 0.95109 ± 0.00092],
#     2.0 => [0.97009 ± 0.00027, 0.95709 ± 0.00035, 0.95099 ± 0.00041, 0.9504 ± 0.00066, 0.9516 ± 0.0019],
#     3.0 => [0.94561 ± 0.00012, 0.95116 ± 0.00021, 0.95442 ± 0.00023, 0.95754 ± 0.00039, 0.96014 ± 0.00054],
#     4.0 => [0.95522 ± 0.00012, 0.96049 ± 0.00023, 0.96212 ± 0.00029, 0.96639 ± 0.00054, 0.968 ± 0.0014],
#     5.0 => [0.95845 ± 0.00029, 0.96925 ± 0.00068, 0.9696 ± 0.001, 0.9752 ± 0.0015, 0.9766 ± 0.0052],
# )

meff_data = Dict(
    # Fixed lambda for rs < 3
    0.5 => [0.94957 ± 0.00012, 0.95751 ± 0.00013, 0.95808 ± 0.00013, 0.95839 ± 0.00016, 0.95861 ± 0.00041],
    1.0 => [0.93574 ± 0.0001, 0.94731 ± 0.00011, 0.94992 ± 0.00013, 0.95129 ± 0.00025, 0.95109 ± 0.00092],
    2.0 => [0.97009 ± 0.00027, 0.95709 ± 0.00035, 0.95099 ± 0.00041, 0.9504 ± 0.00066, 0.9516 ± 0.0019],
    # Minimal sensitivity for rs ≥ 3
    # NOTE: N = 1 has a plateau at λ = ∞ => (m*/m)₁ = 1
    3.0 => [1.0 ± 0.0, 0.95116 ± 0.00021, 0.95442 ± 0.00023, 0.95754 ± 0.00039, 0.96014 ± 0.00054],
    4.0 => [1.0 ± 0.0, 0.96049 ± 0.00023, 0.96212 ± 0.00029, 0.96639 ± 0.00054, 0.968 ± 0.0014],
    5.0 => [1.0 ± 0.0, 0.96925 ± 0.00068, 0.9696 ± 0.001, 0.9752 ± 0.0015, 0.9766 ± 0.0052],
)
lambdas = Dict(
    0.5 => [3.5, 3.5, 3.5, 3.5, 3.5],
    1.0 => [1.75, 1.75, 1.75, 1.75, 1.75],
    2.0 => [2.0, 2.0, 2.0, 2.0, 2.0],
    3.0 => [0.75, 0.75, 1.0, 1.25, 1.75],
    4.0 => [0.625, 0.625, 0.75, 1.0, 1.125],
    5.0 => [0.5, 0.5, 0.625, 0.875, 0.875],
)

style = PyPlot.matplotlib."style"
style.use(["science", "std-colors"])
color = [
    "black",
    cdict["orange"],
    cdict["blue"],
    cdict["cyan"],
    cdict["magenta"],
    cdict["red"],
    # cdict["teal"],
]
rcParams = PyPlot.PyDict(PyPlot.matplotlib."rcParams")
rcParams["font.size"] = 16
rcParams["font.family"] = "Times New Roman"
figure(figsize=(6, 4))

for o in 1:maxOrder
    # for o in maxOrder:maxOrder
    yval = [meff_data[rs][o].val for rs in rslist]
    yerr = [meff_data[rs][o].err for rs in rslist]
    println(rslist)
    println(yval)
    println(yerr)

    # Plot order-by-order lambda optima for rs > 2 at d = 3
    errorbar(
        rslist,
        yval,
        yerr=yerr,
        # color="black",
        color=color[o],
        capsize=4,
        fmt="o",
        markerfacecolor="none",
        label="\$N = $o\$",
        zorder=10 * o,
    )
    if o > 1
        xfit, yfit = spline_with_bc(rslist, yval, yerr)
        # xfit, yfit = spline(rslist, yval, yerr)
        plot(xfit, yfit; color=color[o], linestyle="--")
        # plot(xfit, yfit; color="black", linestyle="--")

        # Estimate the turning point (local minimum) of the mass ratio to max order
        if o == maxOrder
            println("\nTurning point: rs = $(xfit[argmin(yfit)])")
            println("Effective mass ratio at turning point: $(minimum(yfit))")
        end
    else
        axhline(1.0; color=color[o], linestyle="--")
    end
end
text(
    3.5,
    0.94,
    "\$\\beta \\hspace{0.1em} \\epsilon_F = $(beta)\$";
    fontsize=16
)
legend(; loc=(0.1, 0.575), ncol=2)
# xticks(range(0.0, 5.0, step=0.5))
xlim(-0.2, 5.2)
ylim(0.93, 1.005)
xlabel("\$r_s\$")
ylabel("\$m^\\star / m\$")
if spin == 2
    savefig("meff$(dim)d_beta$(beta)_vs_rs.pdf")
else
    savefig("meff$(dim)d_beta$(beta)_spin$(spin)_vs_rs.pdf")
end

# Plot meff convergence at each rs
for rs in rslist
    figure(figsize=(6, 4))
    orders = 1:maxOrder
    yval = Measurements.value.(meff_data[rs])
    yerr = Measurements.uncertainty.(meff_data[rs])
    println(yval)
    println(yerr)

    # Plot order-by-order lambda optima for rs > 2 at d = 3
    errorbar(
        orders,
        yval,
        yerr=yerr,
        color=cdict["blue"],
        capsize=4,
        fmt="o-",
        markerfacecolor="none",
    )
    if rs > 2
        for o in orders
            lambdastr = o == 1 ? "\\infty" : "$(lambdas[rs][o])"
            annotate(
                "\$$lambdastr\$",
                (o, yval[o]),
                xytext=(o + 0.1, yval[o] - 0.0025),
                fontsize=12,
            )
        end
    end

    # xfit, yfit = spline(orders, yval, yerr)
    # plot(xfit, yfit; color=cdict["blue"], linestyle="--")
    lambdastr = rs < 3 ? ",\\, \\lambda^\\star = $(lambdas[rs][1])" : ""
    yloc = 0.99
    if rs == 0.5
        yloc=0.9505
        xlim(0.75, 5.25)
        ylim(0.9485, 0.9605)
    elseif rs == 1.0
        yloc=0.9375
        xlim(0.75, 5.25)
        ylim(0.934, 0.956)
    elseif rs == 2.0
        yloc = 0.9675
        xlim(0.75, 5.25)
        ylim(0.949, 0.971)
    elseif rs == 3.0
        yloc = 0.99
        xlim(0.75, 5.65)
        ylim(0.945, nothing)
    elseif rs == 4.0
        yloc = 0.992
        xlim(0.75, 5.65)
        ylim(0.955, nothing)
    elseif rs == 5.0
        yloc = 0.994
        xlim(0.75, 5.65)
        ylim(0.965, nothing)
    end
    text(
        rs < 3 ? 2.0 : 3.0,
        yloc,
        "\$r_s = $(rs),\\, \\beta \\hspace{0.1em} \\epsilon_F = $(beta)$lambdastr\$";
        fontsize=16
    )
    xlabel("Perturbation order \$N\$")
    ylabel("\$m^\\star / m\$")
    if spin == 2
        savefig("meff$(dim)d_rs=$(rs)_beta$(beta)_vs_N.pdf")
    else
        savefig("meff$(dim)d_rs=$(rs)_beta$(beta)_spin$(spin)_vs_N.pdf")
    end
end

File: example/sigma/plot_meff_vs_rs.py
#!/usr/bin/python3
import scienceplots
import numpy as np
from scipy import interpolate
import matplotlib as mat
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1.inset_locator import inset_axes

cdict = {
    "blue" : "#0077BB",
    "cyan" : "#33BBEE",
    "teal" : "#009988",
    "orange" : "#EE7733",
    "red" : "#CC3311",
    "magenta" : "#EE3377",
    "grey" : "#BBBBBB",
}
plt.switch_backend("TkAgg")
plt.style.use(["science", "std-colors"])
# mat.rcParams.update({'font.size': 28})
# mat.rcParams.update({"font.size": 16})
mat.rcParams["font.size"] = 16
mat.rcParams["mathtext.fontset"] = "cm"
mat.rcParams["font.family"] = "Times New Roman"
# size = 36
# sizein = 12

points = ['o','^','s','v','p','<','h','>']
# Colormap = ["blue", "red", "orange", "violet", "green", "black", "darkblue"]
# Color_beta = {0.125: "red", 0.25: "orange", 0.5: "violet", 1.0: "green", 2.0: "blue", 2.5:"darkblue", 16.0: "black"}
Colormap = [
    cdict["blue"],
    cdict["red"],
    cdict["orange"],
    cdict["magenta"],
    cdict["cyan"],
    "black",
    cdict["teal"],
]
# pt_beta = {0.125: 'o', 0.25: '^', 0.5: 'h', 1.0: 'v', 2.0: 'd', 2.5:'p', 16.0: 's'}
pts = ['s', '^', 'v', 'p', 's', 'o', 'd']
# fig, axes = plt.subplots(1,2, sharex='col')#, sharey=True, gridspec_kw={'wspace': 0})
# fig, ax1 = plt.subplots(figsize=(6, 4.6))
fig, ax1 = plt.subplots(figsize=(6, 4))
# fig.set_size_inches(13,10)

# m_VDMC = [1.0, 0.95861, 0.9511, 0.9516, 0.9601, 0.968, 0.9782]
# m_VDMC_err = [0, 0.00089, 0.0021, 0.0031, 0.0022, 0.0021, 0.0025]
# rs_VDMC = [0, 0.5, 1, 2, 3, 4, 5]
m_VDMC = [1.0, 0.95893, 0.9514, 0.9516, 0.9597, 0.9692, 0.9771, 0.9842]
m_VDMC_err = [0, 0.00067, 0.0016, 0.0018, 0.0016, 0.0026, 0.0028, 0.0029]
rs_VDMC = [0, 0.5, 1, 2, 3, 4, 5, 6]

rs_VMC = [0, 1, 2, 4, 5, 10]
# m_BFVMC = [[1.00, 0.01], [0.98, 0.01], [1.00, 0.02], [1.09,0.03],[1.28, 0.03]]
m_BFVMC = [1.0, 1.00, 0.98, 1.00, 1.09, 1.28]
m_BFVMC_err = [0, 0.01, 0.01, 0.02, 0.03, 0.03]
# m_SJVMC = [[0.96,0.01], [0.94, 0.02], [0.94, 0.02], [1.02, 0.02], [1.13, 0.03]]
m_SJVMC = [1.0, 0.96, 0.94, 0.94, 1.02, 1.13]
m_SJVMC_err = [0, 0.01, 0.02, 0.02, 0.02, 0.03]

rs_DMC = [0, 1, 2, 3, 4, 5]
m_DMC = [1.0, 0.918, 0.879, 0.856, 0.842, 0.791]
m_DMC_err = [0, 0.006, 0.014, 0.014, 0.017, 0.01]

rs_RPA = [0, 1, 2, 3, 4, 5, 6]
# m_RPA = [[0.97, 0], [0.992,0], [1.016,0] [1.039,0], [1.059, 0]]
# m_Gp = [[0.952, 0], [0.951, 0], [0.956, 0], [0.962, 0], [0.968, 0]]
# m_Gpm = [[0.957, 0], [0.966, 0], [0.983, 0], [1.005, 0], [1.028,0]]
m_G0W0 = [1.0, 0.970, 0.992, 1.016, 1.039, 1.059, 1.078]
m_Gp = [1.0, 0.952, 0.951, 0.956, 0.962, 0.968, 0.973]
m_Gpm = [1.0, 0.957, 0.966, 0.983, 1.005, 1.028, 1.055]

def errorbar_mvsrs(rs, mdata, merr, idx, label, ax=ax1, zorder=None):
    if zorder is not None:
        handle = ax.errorbar(rs, mdata, merr, fmt=pts[idx], capthick=1, capsize=4,
                ms=5, color=Colormap[idx], label=label, zorder=zorder)
    else:
        handle = ax.errorbar(rs, mdata, merr, fmt=pts[idx], capthick=1, capsize=4,
                    ms=5, color=Colormap[idx], label=label)
    return handle

def plot_mvsrs(rs, mdata, idx, label, ls='-', ax=ax1):
    # mfitfunc = interpolate.PchipInterpolator(rs, mdata)
    mfitfunc = interpolate.Akima1DInterpolator(rs, mdata)
    # xgrid = np.arange(0, 6.2, 0.02)
    xgrid = np.arange(0, 6.2, 0.02)
    # ax.plot(rs, mdata, 'o', ms=10, color=Colormap[idx])
    # ax.plot(xgrid, mfitfunc(xgrid), ls=ls, lw=2, color=Colormap[idx], label=label)
    handle, = ax.plot(xgrid, mfitfunc(xgrid), ls=ls, color=Colormap[idx], label=label)

    yfit = np.ma.masked_invalid(mfitfunc(xgrid))
    # print(yfit)
    print("Turning point: rs = ", xgrid[np.argmin(yfit)])
    print("Effective mass ratio at turning point: ", np.min(yfit))
    return handle

# m_QMC = [m_DMC, m_SJVMC]
# m_QMC_errs = [m_DMC_err, m_SJVMC_err]
# labels = ["DMC (2021)", "VMC (2023)"]
# idx = 0
# for (mdat, merr, label) in zip(m_QMC, m_QMC_errs, labels):
    # errorbar_mvsrs(rs_VMC, mdat, merr, idx, label)
    # idx = idx + 1
handle1 = errorbar_mvsrs(rs_DMC, m_DMC, m_DMC_err, 0, "DMC [11]", zorder=10)
handle2 = errorbar_mvsrs(rs_VMC, m_SJVMC, m_SJVMC_err, 1, "VMC [12]", zorder=20)

l1_handles = []
m_RPA = [m_G0W0, m_Gp, m_Gpm]
# merr = np.zeros(len(rs_RPA))
labels = [r"$G_0W_0$", r"$G_+$", r"$G_+$\,\&\,$G_-$"]
# labels = [r"$G_0W_0$", r"$G_+$ [15]", r"$G_+$\,\&\,$G_-$ [15]"]
idx = 2
for (mdat, label) in zip(m_RPA, labels):
    print("\nPlotting ", label)
    handle = plot_mvsrs(rs_RPA, mdat, idx, label, '--')
    l1_handles.append(handle)
    idx = idx + 1

handle3 = errorbar_mvsrs(rs_VDMC, m_VDMC, m_VDMC_err, idx, "Our data", zorder=30)
# rs_VDMC.append(5.2)
# m_VDMC.append(0.98)
print("\nPlotting our data")
plot_mvsrs(rs_VDMC, m_VDMC, idx, "", '-')

ax1.set_xlabel(r"$r_s$")
ax1.set_ylabel(r"$m^*/m$")
ax1.set_xlim(0, 6.2)
ax1.set_ylim(0.745, 1.095)
# ax1.set_ylim(0.78, 1.065)
# ax1.set_ylim(0.94, 0.99)
# ax1.legend(loc="upper left", fontsize=14, ncol=2)

# Assemble legends
l2_handles = [handle1, handle2, handle3]
# top_legend = plt.legend(handles=l1_handles, loc="upper left", fontsize=14, title="Ref.[15]")
top_legend = plt.legend(handles=l1_handles, loc="upper left", fontsize=14)
bottom_legend = plt.legend(handles=l2_handles, loc="lower left", fontsize=14)
ax1.add_artist(top_legend)
ax1.add_artist(bottom_legend)
ax1.set_xticks([0, 1, 2, 3, 4, 5, 6])
ax1.set_yticks([0.8, 0.85, 0.9, 0.95, 1.0, 1.05])

# ax1.set_xlabel(r"$r_s$", size=34)
# ax1.set_ylabel(r"$m^*/m$", size=34)
# ax1.tick_params(labelsize=30)
# ax1.legend(loc=2, labelspacing=1)
# ax1.legend(loc=3, fontsize=26)

plt.savefig("meff_3DUEG.pdf")